fwerror/function_error.go

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package fwerror
+
+// FunctionError is an interface for errors generated during the execution
+// of provider-defined functions.
+//
+// See the NewFunctionError constructor for a generic implementation.
+//
+// To add argument position information to an existing function error,
+// see the WithFunctionArgument function.
+type FunctionError interface {
```

Comment:
Given how bespoke these will be compared to other provider development functionality, it may be good to consider:

* Consolidating this into the existing `function` package as it just happens to be a very awkward situation these error types are being introduced, so (famous last words) other error types would not be created in preference of diagnostics, or if they were, they would hopefully be unrelated to functions.
* Whether `FunctionError` (`function.Error` if above happens) should be simplified to a `struct` type instead of an interface. I'm not sure whether we want the extensibility/complexity that is offered by the `diag` package here. It might be interesting to peek in source available provider code whether developers are leveraging the extensibility of diagnostics via those interface(s) or just implementing the existing types/methods. Implementing Go 1.13+ error wrapping capabilities might be good enough, where diagnostics do have separate capabilities beyond pure error handling. 😄

Reply:
I've gone ahead and refactored `fwerror.FunctionErrors` to `function.FuncError`. 

These changes have removed the `fwerror.FunctionError`, and `fwerror.FunctionErrorWithFunctionArgument` interfaces, as well as removing the `fwerror.FunctionErrors` type. In their place is a single `function.FuncError` struct type. 

A `function.ConcatFuncErrors` helper function has been created which concatenates the `function.FuncError` text fields of two or more instances of `function.FuncError`. I thought this might represent the simplest implementation that would allow handling of multiple instances of `function.FuncError`, given that `errors.Join` returns the `error` interface, requiring that we either alter the return type from `function.FuncError` to `error` and use type assertion where necessary, or implement our own version of `Join`, specifically for `function.FuncError`. In the latter case, unless we re-introduce an interface, I believe we'll need to add a `errs []*FuncError` field to `function.FuncError` struct type, which seems a little odd to me.

Interested to hear your thoughts on the `function.FuncError` implementation, and the associated helper functions.

The usages of `github.com/hashicorp/terraform-plugin-framework/diag` that I found all appeared to be using the pre-defined diagnostics rather than new implementations of the available interfaces:

- https://github.com/DataDog/terraform-provider-datadog/blob/46998f7edae69f6ecd01d522fb7c7fb92e7f401c/datadog/internal/utils/utils.go
- https://github.com/elastic/terraform-provider-elasticstack/blob/4d6e2cbc8d6c09eb3c926ed0be864a89017134e7/internal/clients/config/kibana.go
- https://github.com/lxc/terraform-provider-incus/blob/7dba95ee15d906db2489b96ef18343141c0c4a15/internal/common/incus_import.go
- https://github.com/linode/terraform-provider-linode/blob/b4fb82bcd8e20947be488777aef22092315d85f3/linode/firewall/framework_models.go
-  https://github.com/dnsimple/terraform-provider-dnsimple/blob/bc71acb3b1d1ced3521814dd52632f411d828b73/internal/framework/resources/zone_resource.go

Reply:
I read this after I left [my comment](https://github.com/hashicorp/terraform-plugin-framework/pull/925#discussion_r1503342432) lol. TLDR: I like `concat`